creusot/tests/should_succeed/hillel.rs

```diff
@@ -0,0 +1,206 @@
+#![feature(allocator_api)]
+#![allow(dead_code)]
+
+extern crate creusot_contracts;
+use creusot_contracts::*;
+
+#[ensures((@^str).len() >= @len && (@^str).len() >= (@str).len())]
+#[ensures((@^str).len() == @len || (@^str).len() == (@str).len())]
+#[ensures(@len <= (@str).len() ==> (@^str).len() == (@str).len())]
+#[ensures(@len > (@str).len() ==> (@^str).len() == @len)]
+#[ensures(forall<i: Int> 0 <= i && i < (@str).len() ==> (@^str)[i] == (@str)[i])]
+#[ensures(forall<i: Int> (@str).len() <= i && i < @len ==> (@^str)[i] == pad)]
+fn right_pad<T: Copy>(str: &mut Vec<T>, len: usize, pad: T) {
+    let old_str = ghost! { str };
+
+    #[invariant(proph_const, ^str == ^*old_str)]
+    #[invariant(old_str_bound, (@old_str).len() <= (@str).len())]
+    #[invariant(len_bound, (@old_str).len() < @len ==> (@str).len() <= @len)]
+    #[invariant(len_bound, (@str).len() > @len ==> (@str).len() == (@old_str).len())]
+    #[invariant(old_elem, forall<i: Int> 0 <= i && i < (@old_str).len() ==> (@str)[i] == (@old_str)[i])]
+    #[invariant(pad_elem, forall<i: Int> (@old_str).len() <= i && i < (@str).len() ==> (@str)[i] == pad)]
+    while str.len() < len {
+        str.push(pad);
+    }
+}
+
+#[ensures((@^str).len() >= @len && (@^str).len() >= (@str).len())]
+#[ensures((@^str).len() == @len || (@^str).len() == (@str).len())]
+#[ensures(forall<i: Int> 0 <= i && i < ((@^str).len() - (@str).len()) ==> (@^str)[i] == pad)]
+#[ensures(forall<i: Int> 0 <= i && i < (@str).len() ==> (@^str)[i + ((@^str).len() - (@str).len())] == (@str)[i])]
+fn left_pad<T: Copy>(str: &mut Vec<T>, len: usize, pad: T) {
+    let old_str = ghost! { str };
+    let mut c: Ghost<usize> = ghost! { 0 };
+
+    #[invariant(proph_const, ^str == ^*old_str)]
+    #[invariant(old_str_bound, (@old_str).len() <= (@str).len())]
+    #[invariant(len_bound, (@old_str).len() < @len ==> (@str).len() <= @len)]
+    #[invariant(len_bound, (@str).len() > @len ==> (@str).len() == (@old_str).len())]
+    #[invariant(count, @c == (@str).len() - (@old_str).len())]
+    #[invariant(old_elem, forall<i: Int> @c <= i && i < (@str).len() ==> (@str)[i] == (@old_str)[i - @c])]
+    #[invariant(pad_elem, forall<i: Int> 0 <= i && i < @c ==> (@str)[i] == pad)]
+    while str.len() < len {
+        str.insert(0, pad);
+        c = ghost! { (1 + c.inner()) };
+    }
+}
+
+#[predicate]
+fn is_unique<T: Eq + Model>(s: Seq<T>) -> bool {
+    pearlite! {
+        forall<i: Int, j :Int> 0 <= i && i < s.len() && 0 <= j && j < s.len() ==> @(s[i]) == @(s[j]) ==> i == j
+    }
+}
+
+#[predicate]
+fn contains<T: Eq + Model>(seq: Seq<T>, elem: T) -> bool {
```

Comment:
Exists in `Seq` now.

Reply:
Good to know! However, there is a slight difference between the two functions: my contains checks for equality of the elements' models while the one in `Seq` directly compares the elements. WIth the new contains function, the [assertion](https://github.com/xldenis/creusot/pull/526/files#diff-62b5c0619a5e5b84a19bf7edfeeb213af2ef0ddb2e67817cfbac6e1870853eafR87) `proof_assert! { contains(@vec, elem) };` no longer holds after checking `vec[i] == elem`.

Reply:
Ah yea, that's a good point, god models are such a pain sometimes...